### PR TITLE
Adding computerFileSize in OC.Util

### PR DESCRIPTION
**Pull request by ppaysant**
## Description
Add a computerFileSize function in js OC.Util.

## Related Issue
#26788

## Motivation and Context
This function may help. For instance, I'm calling the setquota.php in settings/ajax/ from one of my app and the quota size returned is in human readable format. I need the size in bytes.

## How Has This Been Tested?
Unit tested in coreSpec.js

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [X] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have added tests to cover my changes.
- [X] All new and existing tests passed.

### CODE REVIEW

File: core/js/js.js [hunk 1]
```diff
@@ -1684,6 +1684,51 @@ function humanFileSize(size, skipSmallSizes) {
 }
 
 /**
+ * Returns a file size in bytes from a humanly readable string
+ * @param  {string} string file size in human readable format
+ * @return {number}
```
Comment (PVince81) [inline core/js/js.js]: returns the parsed number or null if the string could not be parsed

File: core/js/js.js [hunk 2]
```diff
@@ -1684,6 +1684,51 @@ function humanFileSize(size, skipSmallSizes) {
...
+		var matches = s.match(/([kmgtp]?b?)$/i);
+		if (matches[1]) {
+			bytes = bytes * bytes_array[matches[1]];
+		}
+		else {
+			return false;
+		}
+
+		bytes = Math.round(bytes);
+		console.log(bytes);
```
Comment (PVince81) [inline core/js/js.js]: remove logging

File: core/js/js.js [hunk 3]
```diff
@@ -1724,6 +1769,7 @@ function relative_modified_date(timestamp) {
 OC.Util = {
 	// TODO: remove original functions from global namespace
 	humanFileSize: humanFileSize,
+	computerFileSize: computerFileSize,
```
Comment (PVince81) [inline core/js/js.js]: please add your function directly here.

The reason `humanFileSize` is separate is because it used to be global in the past, but global is bad... so we added a shortcut here.

However `computerFileSize` never existed before, so we have a chance to have it directly in the namespace and no global.

File: core/js/tests/specs/coreSpec.js [hunk 4]
```diff
@@ -590,6 +590,26 @@ describe('Core base tests', function() {
...
 		});
+		describe('computerFileSize', function() {
+			it('correctly parses file sizes from a human readable formated string', function() {
+				var data = [
+					['0 B', 0],
+					['125 B', 125],
+					['125b', 125],
+					['125 KB', 128000],
+					['125kb', 128000],
+					['122.1 MB', 128031130, ],
```
Comment (PVince81) [inline core/js/tests/specs/coreSpec.js]: trailing comma in array ?

File: core/js/tests/specs/coreSpec.js [hunk 5]
```diff
@@ -590,6 +590,26 @@ describe('Core base tests', function() {
 				}
 			});
 		});
+		describe('computerFileSize', function() {
+			it('correctly parses file sizes from a human readable formated string', function() {
```
Comment (PVince81) [inline core/js/tests/specs/coreSpec.js]: add a test for failed parse and check that it returns `null`

**Comment by PVince81** [comment]
Looks good so far, and now I realize I missed an opportunity to properly use Github's "Review" feature 😕

File: core/js/js.js [hunk 6]
```diff
@@ -1726,6 +1726,50 @@ OC.Util = {
...
+	 * Inspired by computerFileSize in helper.php
+	 */
+	computerFileSize: function (string) {
+		var s = string.toLowerCase();
+
+		if (!isNaN(parseFloat(s)) && isFinite(s)) {
+			return parseFloat(s);
+		}
+
+		var bytes_array = {
```
Comment (PVince81) [inline core/js/js.js]: use camel case `bytesArray`

File: core/js/js.js [hunk 7]
```diff
@@ -1726,6 +1726,50 @@ OC.Util = {
...
+			'pb': 1024 * 1024 * 1024 * 1024 * 1024,
+			'p' : 1024 * 1024 * 1024 * 1024 * 1024
+		};
+
+		var bytes = parseFloat(s);
+
+		var matches = s.match(/([kmgtp]?b?)$/i);
+		if (matches[1]) {
+			bytes = bytes * bytes_array[matches[1]];
+		}
```
Comment (PVince81) [inline core/js/js.js]: `} else {` looks better

File: core/js/js.js [hunk 8]
```diff
@@ -1726,6 +1726,50 @@ OC.Util = {
 	humanFileSize: humanFileSize,
 
 	/**
+	 * Returns a file size in bytes from a humanly readable string
+	 * @param  {string} string file size in human readable format
+	 * @return {number} or null if string could not be parsed
+	 *
+	 * Makes 2kB to 2048.
+	 *
+	 * Inspired by computerFileSize in helper.php
```
Comment (PVince81) [inline core/js/js.js]: Move these comments in the header, under `Returns a file size...`

File: core/js/js.js [hunk 9]
```diff
@@ -1726,6 +1726,50 @@ OC.Util = {
...
+			'm' : 1024 * 1024,
+			'gb': 1024 * 1024 * 1024,
+			'g' : 1024 * 1024 * 1024,
+			'tb': 1024 * 1024 * 1024 * 1024,
+			't' : 1024 * 1024 * 1024 * 1024,
+			'pb': 1024 * 1024 * 1024 * 1024 * 1024,
+			'p' : 1024 * 1024 * 1024 * 1024 * 1024
+		};
+
+		var bytes = parseFloat(s);
```
Comment (PVince81) [inline core/js/js.js]: I noticed that you're calling `parseFloat()` three times in this method.
Would be better to just call it once at the top `var bytes = parseFloat(s);`, then do the `isNan()` tests, etc

File: core/js/tests/specs/coreSpec.js [hunk 10]
```diff
@@ -590,6 +590,27 @@ describe('Core base tests', function() {
...
+					['125 B', 125],
+					['125b', 125],
+					['125 KB', 128000],
+					['125kb', 128000],
+					['122.1 MB', 128031130],
+					['122.1mb', 128031130],
+					['119.2 GB', 127990025421],
+					['119.2gb', 127990025421],
+					['116.4 TB', 127983153473126],
+					['116.4tb', 127983153473126],
```
Comment (PVince81) [inline core/js/tests/specs/coreSpec.js]: also add a case to cover the `!isNan()` + `isFinite()` case.

And add a test for pure numbers, without unit.

**Review by PVince81 — COMMENTED** [review]
Second level review in sharp mode, stuff I missed the first time.

File: core/js/js.js [hunk 11]
```diff
@@ -1726,6 +1726,52 @@ OC.Util = {
...
+	 */
+	computerFileSize: function (string) {
+		if (typeof string != 'string') {
+			return null;
+		}
+
+		var s = string.toLowerCase();
+		var bytes = parseFloat(s)
+
+		if (!isNaN(bytes) && isFinite(s)) {
```
Comment (PVince81) [inline core/js/js.js]: I guess `isFinite` does not work with `bytes` directly ? Just asking, am not familiar with that method

**Comment by PVince81** [comment]
Looks good 👍
